lib/config.js

```diff
@@ -212,127 +87,277 @@ class Config {
-                parser: this.parser
-            });
-        }
+            // Check for root flag
+            if (localConfig.root === true) {
+                rootPath = path.dirname(localConfigFile);
+            }
 
-        if (this.parserOptions) {
-            config = ConfigOps.merge(config, {
-                parserOptions: this.parserOptions
-            });
+            debug(`Using ${localConfigFile}`);
+            configs.push(localConfig);
+            searched.push(localConfigDirectory);
         }
 
-        // Step 8: Merge in command line plugins
-        if (this.options.plugins) {
-            debug("Merging command line plugins");
-            Plugins.loadAll(this.options.plugins);
-            config = ConfigOps.merge(config, { plugins: this.options.plugins });
-        }
+        if (!configs.length && !cache && !this.useSpecificConfig) {
 
-        // Step 9: Apply environments to the config if present
-        if (config.env) {
-            config = ConfigOps.applyEnvironments(config);
+            // Fall back on the personal config from ~/.eslintrc
+            debug("Using personal config file");
+            const personalConfig = this.getPersonalConfig();
+
+            if (personalConfig) {
+                configs.push(personalConfig);
+            } else if (!hasRules(this.options) && !this.options.baseConfig) {
+
+                // No config file, no manual configuration, and no rules, so error.
+                const noConfigError = new Error("No ESLint configuration found.");
+
+                noConfigError.messageTemplate = "no-config-found";
+                noConfigError.messageData = {
+                    directory,
+                    filesExamined: localConfigFiles
+                };
+
+                throw noConfigError;
+            }
         }
 
-        this.cache[directory] = config;
+        // Merged with any cached portion
+        configs.reverse();
+        cache = cache ? cache.concat(configs) : configs;
 
-        return config;
+        // Set the caches for the parent directories
+        configCache.setHierarchyLocalConfigs(searched, cache);
+
+        return cache;
     }
 
     /**
-     * Find local config files from directory and parent directories.
+     * Gets the vector of applicable configs and subconfigs from the hierarchy for a given file. A vector is an array of
+     * objects, each containing a config file path and an array of override indices corresponding to entries in config
+     * file's overrides section whose glob patterns match the specified file path; e.g., the vector
+     * { configFile: '/home/john/app/.eslintrc', matchingOverrides: [0, 2] } would indicate that the main project
+     * .eslintrc file and its first and third override blocks apply to the current file.
+     * @param {string} filePath The file path for which to build the hierarchy and config vector.
+     * @returns {Array<Object>} config vector applicable to the specified path
```

Comment:
It's an interesting idea, but I'm not sure how exactly it would work. Right now, the vector is used as a sort of unique identifier that can be hashed to fetch a merged config object for a file. The logic is: take the file's path, match it against the override patterns from all the configs above it in the hierarchy to get a vector, then hash the vector to see whether there exists a merged config for that exact combination of configs + overrides. If the vector contained the overrides themselves, we'd still need a way to figure out whether there's a merged config in the cache for that file, otherwise we'd have to do the work of merging the overrides for every file. So I don't think it would gain us anything, unless I'm misunderstanding what you're proposing.